external-crates/move/crates/move-vm-runtime/src/jit/execution/ast.rs

```diff
@@ -1036,13 +966,57 @@ impl ArenaType {
             ArenaType::Datatype(def_idx) => Type::Datatype(def_idx.clone()),
             ArenaType::DatatypeInstantiation(def_inst) => {
                 let (def_idx, instantiation) = &**def_inst;
-                let inst = instantiation.into_iter().map(|ty| ty.to_type()).collect();
+                let inst = instantiation.iter().map(|ty| ty.to_type()).collect();
                 Type::DatatypeInstantiation(Box::new((def_idx.clone(), inst)))
             }
         }
     }
 }
 
+impl DatatypeDescriptor {
+    pub fn new(
+        name: Identifier,
+        defining_id: ModuleId,
+        runtime_id: ModuleId,
+        datatype_info: ArenaBox<Datatype>,
+    ) -> Self {
+        Self {
+            name,
+            defining_id,
+            runtime_id,
+            datatype_info,
+        }
+    }
+
+    pub fn type_parameters(&self) -> &[DatatypeTyParameter] {
+        match self.datatype_info.inner_ref() {
+            Datatype::Enum(vmpointer) => &vmpointer.type_parameters,
+            Datatype::Struct(vmpointer) => &vmpointer.type_parameters,
+        }
+    }
+
+    pub fn abilities(&self) -> &AbilitySet {
+        match self.datatype_info.inner_ref() {
+            Datatype::Enum(vmpointer) => &vmpointer.abilities,
+            Datatype::Struct(vmpointer) => &vmpointer.abilities,
+        }
+    }
+
+    pub fn qualified_name(&self) -> VirtualTableKey {
+        match self.datatype_info.inner_ref() {
+            Datatype::Enum(ptr) => ptr.to_ref().def_vtable_key.clone(),
+            Datatype::Struct(ptr) => ptr.to_ref().def_vtable_key.clone(),
+        }
+    }
+
+    pub fn intra_package_name(&self) -> IntraPackageKey {
+        match self.datatype_info.inner_ref() {
+            Datatype::Enum(ptr) => ptr.def_vtable_key.inner_pkg_key,
+            Datatype::Struct(ptr) => ptr.def_vtable_key.inner_pkg_key,
+        }
```

Comment:
Consider rewriting in terms of `qualified_name` possibly

Reply:
That would technically clone the whole key instead of just the inner key, which is a much larger copy.